hail/python/test/hailtop/aiotools/test_copy.py

```diff
@@ -416,3 +417,135 @@ async def test_copy_src_parts(copy_test_context):
 
     await expect_file(fs, f'{dest_base}file1', 'src/a/file1')
     await expect_file(fs, f'{dest_base}subdir/file2', 'src/a/subdir/file2')
+
+
+async def write_file(fs, url, data):
+    async with await fs.create(url) as f:
+        await f.write(data)
+
+
+@pytest.mark.asyncio
+async def test_file_and_directory_error_with_slash_empty_file(router_filesystem):
+    sema, fs, bases = router_filesystem
+
+    src_base = await fresh_dir(fs, bases, 'gs')
+
+    await write_file(fs, f'{src_base}empty/', '')
+    await write_file(fs, f'{src_base}empty/foo', b'foo')
+
+    await fs.listfiles(f'{src_base}')
+    await fs.listfiles(f'{src_base}', recursive=True)
+    await fs.listfiles(f'{src_base}empty/')
+    await fs.listfiles(f'{src_base}empty/', recursive=True)
```

Comment:
`listfiles` returns an async generator, and you won't test the full behavior unless you drain the generator.